File: files/Python/fileclient.py
"""
Copyright Core DF

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

Local file and SFTP helpers for Core Auto step scripts.
"""

import os
import shutil
from lib.result import transport_error


def LocalRead(path: str, encoding: str = "utf-8") -> dict:
    try:
        with open(path, "r", encoding=encoding) as fh:
            return {"status_code": 200, "content": fh.read()}
    except OSError as exc:
        return {"status_code": 500, "error": str(exc)}


def LocalWrite(path: str, content: str, encoding: str = "utf-8") -> dict:
    try:
        parent = os.path.dirname(path)
        if parent:
            os.makedirs(parent, exist_ok=True)
        with open(path, "w", encoding=encoding) as fh:
            fh.write(content)
        return {"status_code": 200}
    except OSError as exc:
        return {"status_code": 500, "error": str(exc)}


def LocalMove(src: str, dest: str) -> dict:
    try:
        shutil.move(src, dest)
        return {"status_code": 200}
    except OSError as exc:
        return {"status_code": 500, "error": str(exc)}


def _sftp_connect():
    import paramiko

    host = os.environ.get("SFTP_HOST", "")
    user = os.environ.get("SFTP_USER", "")
    password = os.environ.get("SFTP_PASSWORD", "")
    port = int(os.environ.get("SFTP_PORT", "22"))
    key_path = os.environ.get("SFTP_PRIVATE_KEY", "")

    if not host or not user:
        raise ValueError("SFTP_HOST and SFTP_USER required")

    client = paramiko.SSHClient()
    client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
    if key_path:
        key = paramiko.RSAKey.from_private_key_file(key_path)
        client.connect(host, port=port, username=user, pkey=key, timeout=60)
    else:
        if not password:
            raise ValueError("SFTP_PASSWORD or SFTP_PRIVATE_KEY required")
        client.connect(host, port=port, username=user, password=password, timeout=60)
    return client, client.open_sftp()


def SftpGet(remote_path: str, local_path: str) -> dict:
    try:
        ssh, sftp = _sftp_connect()
        try:
            parent = os.path.dirname(local_path)
            if parent:
                os.makedirs(parent, exist_ok=True)
            sftp.get(remote_path, local_path)
        finally:
            sftp.close()
            ssh.close()
        return {"status_code": 200}
    except Exception as exc:
        return transport_error(str(exc))


def SftpPut(local_path: str, remote_path: str) -> dict:
    try:
        ssh, sftp = _sftp_connect()
        try:
            sftp.put(local_path, remote_path)
        finally:
            sftp.close()
            ssh.close()
        return {"status_code": 200}
    except Exception as exc:
        return transport_error(str(exc))


def SftpList(remote_dir: str = ".") -> dict:
    try:
        ssh, sftp = _sftp_connect()
        try:
            names = sftp.listdir(remote_dir)
        finally:
            sftp.close()
            ssh.close()
        return {"status_code": 200, "files": names}
    except Exception as exc:
        return transport_error(str(exc))

File: files/Node/fileclient.js
// Copyright Core DF
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Local file and SFTP helpers for Core Auto step scripts.

import fs from 'node:fs/promises';
import path from 'node:path';
import { missingEnv, transportError } from './lib/result.js';

async function sftpConnect() {
  let SftpClient;
  try {
    ({ default: SftpClient } = await import('ssh2-sftp-client'));
  } catch {
    return { error: { status_code: 500, error: 'ssh2-sftp-client package required for SFTP' } };
  }

  const host = process.env.SFTP_HOST || '';
  const user = process.env.SFTP_USER || '';
  const password = process.env.SFTP_PASSWORD || '';
  const port = parseInt(process.env.SFTP_PORT || '22', 10);
  const keyPath = process.env.SFTP_PRIVATE_KEY || '';

  if (!host || !user) {
    return { error: missingEnv('SFTP_HOST and SFTP_USER') };
  }

  const sftp = new SftpClient();
  const config = { host, port, username: user };
  if (keyPath) {
    config.privateKey = await fs.readFile(keyPath);
  } else if (password) {
    config.password = password;
  } else {
    return { error: missingEnv('SFTP_PASSWORD or SFTP_PRIVATE_KEY') };
  }

  try {
    await sftp.connect(config);
    return { sftp };
  } catch (exc) {
    return { error: transportError(String(exc.message || exc)) };
  }
}

export async function LocalRead(filePath, encoding = 'utf-8') {
  try {
    const content = await fs.readFile(filePath, encoding);
    return { status_code: 200, content };
  } catch (exc) {
    return { status_code: 500, error: String(exc.message) };
  }
}

export async function LocalWrite(filePath, content, encoding = 'utf-8') {
  try {
    await fs.mkdir(path.dirname(filePath), { recursive: true });
    await fs.writeFile(filePath, content, encoding);
    return { status_code: 200 };
  } catch (exc) {
    return { status_code: 500, error: String(exc.message) };
  }
}

export async function LocalMove(src, dest) {
  try {
    await fs.rename(src, dest);
    return { status_code: 200 };
  } catch (exc) {
    return { status_code: 500, error: String(exc.message) };
  }
}

export async function SftpGet(remotePath, localPath) {
  const conn = await sftpConnect();
  if (conn.error) return conn.error;
  try {
    await fs.mkdir(path.dirname(localPath), { recursive: true });
    await conn.sftp.fastGet(remotePath, localPath);
    await conn.sftp.end();
    return { status_code: 200 };
  } catch (exc) {
    await conn.sftp.end().catch(() => {});
    return transportError(String(exc.message || exc));
  }
}

export async function SftpPut(localPath, remotePath) {
  const conn = await sftpConnect();
  if (conn.error) return conn.error;
  try {
    await conn.sftp.fastPut(localPath, remotePath);
    await conn.sftp.end();
    return { status_code: 200 };
  } catch (exc) {
    await conn.sftp.end().catch(() => {});
    return transportError(String(exc.message || exc));
  }
}

export async function SftpList(remoteDir = '.') {
  const conn = await sftpConnect();
  if (conn.error) return conn.error;
  try {
    const files = await conn.sftp.list(remoteDir);
    await conn.sftp.end();
    return { status_code: 200, files: files.map((f) => f.name) };
  } catch (exc) {
    await conn.sftp.end().catch(() => {});
    return transportError(String(exc.message || exc));
  }
}
